lib/withOnyx.js

```diff
@@ -121,14 +121,9 @@ export default function (mapOnyxToState) {
                     return;
                 }
 
-                const stateUpdate = {...this.tempState, loading: false};
-
-                // The state is set here manually, instead of using setState because setState is an async operation, meaning it might execute on the next tick,
-                // or at a later point in the microtask queue. That can lead to a race condition where
-                // setWithOnyxState is called before the state is actually set. This results in unreliable behavior when checking the loading state and has been mainly observed on fabric.
-                this.state = stateUpdate;
-
-                this.setState(stateUpdate); // Trigger a render
+                // Retain previous state to avoid data loss during pre-load updates.
+                // This handles race conditions where setWithOnyxState might be called after setState
+                this.setState(prevState => ({...this.tempState, ...prevState, loading: false}));
```

Comment:
What is the race condition exactly? Is there any example? Can we break this problem down a bit further in the comment? Who is calling `setState()` and why should we ever prefer the previous state of the `withOnyx()` component over the initial value being set here?

Reply:
Exact flow described [here](https://github.com/Expensify/App/issues/21881#issuecomment-1643878246)

Reply:
That's pretty interesting.

So ultimately, sounds like we have made the bad assumption here that `setState()` can never be called before `setWithOnyxState()`.

If we got here and the `state` already has the key then we could 
 in theory remove it from `tempState`? That approach might be a little more obvious vs. spreading the `prevState` after the `tempState`. Though they would achieve the same result. Thoughts?

Reply:
Yeah, I agree with you that it would be more straightforward to remove properties from the  `tempState` that already exist in `prevState`

I'd assume that you mean something like `omit`

I will  recheck the flow again with the update and apply changes

Reply:
@marcaaron PR updated with omit approach